### PR TITLE
Omk 11098

### PR DESCRIPTION
**Pull request by Sheetalfw**

### CODE REVIEW

File: lib/NMISNG/Auth.pm [hunk 1]
```diff
@@ -549,53 +543,199 @@ sub _file_verify {
...
+		$status = $ldap->bind( $auth_ldap_acc, password => $auth_ldap_psw);
+		# if full debugging dumps are requested, put it in a separate log file
+		if ($auth_ldap_debug)
+		{
+			open(F, ">>", $self->{config}->{'<nmis_logs>'}."/auth-ldap-debug.log");
+			print F NMISNG::Util::returnDateStamp() . ": " . "\$ldap->bind($auth_ldap_acc, password=>**************, version => 3)\n";
+			print F NMISNG::Util::returnDateStamp() . ": " . Dumper($status) ."\n";
+			close(F);
+		}
+		if ($status->code() ne 0) {
```
Comment (markdueck1) [inline lib/NMISNG/Auth.pm]: ne here seems wrong

File: lib/NMISNG/Auth.pm [hunk 2]
```diff
@@ -549,53 +543,199 @@ sub _file_verify {
...
+		if($@ || !$ldap) {
+			NMISNG::Util::logAuth("Auth::_ldap_verify, ERROR: Could not create LDAP session; ERROR: $@");
+			return 0;
 		}
+
+		# check user
+	
+		$status = $ldap->bind("$dn",password=>"$p");
+		NMISNG::Util::logAuth("Auth::_ldap_verify, DEBUG: LDAP bind dn '$d' status ".$status->code()) if ($self->{debug});
+		if ($status->code eq 0) {
```
Comment (markdueck1) [inline lib/NMISNG/Auth.pm]: eq here seems wrong, == ?

File: lib/NMISNG/Auth.pm [hunk 3]
```diff
@@ -2141,4 +2199,383 @@ sub update_last_login
...
+# Get groups for the logged in user from ldap
+# And map the groups with a local privilege from nmis
+# Returns privilege, and group
+sub _get_ldap_privs
+{
+	my ($self, %args) = @_;
+	my $user = $args{"user"};
+	my $ldap;
+    my $sec = 0;
+	my $auth_ldap_base       = $self->{config}->{'auth_ldap_base'} // $self->{config}->{'auth_ldap_context'}; # retured value
```
Comment (markdueck1) [inline lib/NMISNG/Auth.pm]: if this is the same as the code above please move them into a common function

File: lib/NMISNG/Auth.pm [hunk 4]
```diff
@@ -2141,4 +2199,383 @@ sub update_last_login
...
+		$auth_ldap_server     = $self->{config}->{'auth_ms_ldap_server'} if ($self->{config}->{'auth_ms_ldap_server'});
+		$auth_ldaps_server    = $self->{config}->{'auth_ms_ldaps_server'} if ($self->{config}->{'auth_ms_ldaps_server'});
+	}
+
+    if ((!$auth_ldaps_server or $auth_ldaps_server eq "") and (!$auth_ldap_server or $auth_ldap_server eq "")) {
+        NMISNG::Util::logAuth("ERROR Auth::_get_ldap_privs, called but not configured");
+		return 0;
+    }
+
+    if ($auth_ldaps_server and $auth_ldaps_server ne "") {
```
Comment (markdueck1) [inline lib/NMISNG/Auth.pm]: this seems to be checking the same thing?

**Review by markdueck1 — CHANGES_REQUESTED** [review]

File: lib/NMISNG/Auth.pm (hunk 2 above)
Comment (Sheetalfw) [reply lib/NMISNG/Auth.pm]: Resolved

File: lib/NMISNG/Auth.pm (hunk 1 above)
Comment (Sheetalfw) [reply lib/NMISNG/Auth.pm]: Resolved

File: lib/NMISNG/Auth.pm (hunk 3 above)
Comment (Sheetalfw) [reply lib/NMISNG/Auth.pm]: Resolved

File: lib/NMISNG/Auth.pm (hunk 4 above)
Comment (Sheetalfw) [reply lib/NMISNG/Auth.pm]: Resolved

File: lib/NMISNG/Auth.pm [hunk 5]
```diff
@@ -549,53 +543,199 @@ sub _file_verify {
...
 
-	foreach $context ( split ":", $self->{config}->{'auth_ldap_context'}  ) {
-		foreach $attr ( @attrlist ) {
-			$dn = "$attr=$u,".$context;
-			$msg = $ldap->bind($dn, password=>$p) ;
-			if(!$msg->is_error) {
-				$ldap->unbind();
-				return 1;
+	my @attrlist = ( 'uid','cn','sAMAccountName' );
+	@attrlist = split( "[ ,]", $auth_ldap_attr ) if( $auth_ldap_attr );
```
Comment (markdueck1) [inline lib/NMISNG/Auth.pm]: the brackets in this split do not make sense

File: lib/NMISNG/Auth.pm [hunk 6]
```diff
@@ -729,7 +702,7 @@ sub _ldap_verify
 	
 		$status = $ldap->bind("$dn",password=>"$p");
 		NMISNG::Util::logAuth("Auth::_ldap_verify, DEBUG: LDAP bind dn '$d' status ".$status->code()) if ($self->{debug});
-		if ($status->code eq 0) {
+		if (defined $status->code && $status->code == 0) {
```
Comment (markdueck1) [inline lib/NMISNG/Auth.pm:705]: It's a bit nicer to call the function once when doing this ...

example code from net::ldap docs looks like this:
$mesg->code  and  die $mesg->error;          # check for errors

**Review by markdueck1 — APPROVED** [review]

**Pull request merged**